Allow for Modification of Default HTTP Response Headers

Jitsusama opened the pull request:
As per issue #243, this PR allows a user to programmatically, or, through the passing of CLI arguments, modify the default HTTP response headers that uvicorn sends out. In the case of the `Server` header, user specified values will over-ride the default value of _uvicorn_. If multiple `Server` header values are passed, then they will all be kept and will over-ride the default value as well.

The progammatic method of specifying default server headers is to pass a list of tuples to the `Config` class instantiation under the kwarg `headers` like so:

```python
config = Config(
    app=App, headers=[('Server', 'Cool Name'), ('X-Candy', 'Is Yummy')])
```

This also introduces a new CLI argument named `--header`. This argument takes a single argument, which is a header name/value pair delineated by a `:`. IE: `uvicorn --header Server:Coolness`. This argument can be passed multiple times to supply multiple over-riding headers.

Commit:
Write tests for default HTTP response headers mods

Commit:
Allow over-riding of default HTTP response headers via configuration

Commit:
Allow custom default headers to be specified via CLI arguments

Jitsusama commented:
FYI; it looks like the failure is related to black. I’m running Python 3.7, and as per your test script, there is apparently an issue with it and Python 3.7 so I didn’t try to run it. If there are any formatting changes that you would like for me to make, just let me know.

davidism commented:
Is there a way to unset the default `Server` header so it's not sent at all? It's one way to fingerprint a server, which can be considered bad.

Jitsusama commented:
@davidism; not with these changes. I guess for that functionality, you’d probably want middleware to remove it at the last possible second in case the application added one?

If you’re more worried about just removing the default one, I would think maybe adding another config kwarg to specify just removing the default value of Server (however it is specified) could be an option. I purposefully didn’t make the new kwarg over-write all truly default values, as the Date header is quite useful and can’t really be passed as a kwarg due to its dynamic nature. I could change the functionality so that if the kwarg is passed, everything but Date is overridden.

@tomchristie, what do you think?

lovelydinosaur commented:
> Is there a way to unset the default Server header so it's not sent at all? It's one way to fingerprint a server, which can be considered bad.

We could consider removing it by default, or providing a way to remove it, tho let's keep that seperate from this issue at the moment. Welcome to open an issue for it and we can follow up in due course.

lovelydinosaur commented:
Need to run `scripts/lint` to resolve the linting errors: https://travis-ci.org/encode/uvicorn/jobs/503924223

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -182,6 +182,12 @@
     show_default=True,
 )
+@click.option(
+    "--default-header",
```
Just `--header` would be fine.

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -210,6 +216,7 @@ def main(
     ssl_cert_reqs: int,
     ssl_ca_certs: str,
     ssl_ciphers: str,
+    custom_headers: typing.List[str],
```
Just `headers`

lovelydinosaur commented on uvicorn/config.py:
```diff
@@ -92,6 +92,7 @@ def __init__(
         ssl_cert_reqs=ssl.CERT_NONE,
         ssl_ca_certs=None,
         ssl_ciphers="TLSv1",
+        custom_headers=None,
```
Again, let's just go with `headers` here.

lovelydinosaur commented on uvicorn/config.py:
```diff
@@ -140,6 +141,11 @@ def __init__(
+        if custom_headers is None:
+            self.custom_headers = []
+        else:
+            self.custom_headers = custom_headers
```
I'd suggest that in the config, we store the headers as lowercased byte-pairs. It's more efficient for us to do that conversion here, just the once.

I guess we want something like this?...

```python
if headers is None:
    headers = []
self.headers = [
    (key.lower().encode('latin-1'), value.encode('latin-1'))
    for key, value in headers
]
if b'server' not in dict(self.headers):
    self.headers = [(b'server', b'uvicorn')] + self.headers
```

lovelydinosaur replied:
Don't worry about `date` - it doesn't make sense for users to specify that, and if they want to do so. There's plenty of header values that it'd be erronous for users to setup here - if they're setting custom headers it's up to them to do so sensibly.

Jitsusama replied:
I'm always reticent to store much business logic in `__init__`. The stuff I put in there already was because I couldn't think of another way to handle it. But, if you'd like it that way, I can accommodate.

Jitsusama replied:
I found a compromise. I'm going to perform the logic in `config.load()`.

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -372,12 +381,7 @@ def run(self, sockets=None, shutdown_servers=True):
-            self.server_state.default_headers = [
-                (b"server", b"uvicorn"),
-                (b"date", current_date),
-            ]
```
If we're setting up the header values in the config, then I think we end up with something pretty simple here, eg?...


```python 
current_time = time.time()
current_date = formatdate(current_time, usegmt=True).encode()
self.server_state.default_headers = [b"date", current_date)] + self.config.headers
```

Commit:
Apply changes introduced by black linter

Commit:
Remove test for over-riding date header

@tomchristie requested to not worry about over-riding the date header,
so there is no need for this test anymore.

Commit:
Rename kwarg/cli arguments and move header logic from Server to Config

As requested in PR #321, I'm renaming the --default-header CLI argument
to --header, and renaming the kwarg/instance variable custom_headers to
headers. Also, I'm moving (most) of the header generation logic from
main.Server to config.Config, so that most of the headers only get
generated once, instead of on every tick. The most computationally
intense generation will still happen ever tick, and that is the
date-stamp generation.

Jitsusama commented:
Hello @tomchristie. This should address all of the changes you requested. I do have a thought though; why don't we move all header generation into config and store it as a property method? We can still do the custom header merge once upon configuration loading, but we could move the date-stamp generation into the property logic so it is completely removed from main.Server. Does that sound good?

lovelydinosaur commented on uvicorn/config.py:
```diff
@@ -122,6 +124,8 @@ def __init__(
         self.ssl_ca_certs = ssl_ca_certs
         self.ssl_ciphers = ssl_ciphers
+        self._headers = headers if headers else []
+        self.headers = None  # type: List[Tuple[bytes, bytes]]
```
Thinking about this once more - How about we go for `.headers` as the string pairs, and `.encoded_headers` for the byte-encoded pairs?

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -373,11 +381,10 @@ def run(self, sockets=None, shutdown_servers=True):
         if counter % 10 == 0:
             current_time = time.time()
-            current_date = formatdate(current_time, usegmt=True).encode()
+            current_date = formatdate(current_time, usegmt=True)
```
Let's leave this line as it currently is - always best to try to keep the footprint of pull requests as strictly minimal as possible.

Commit:
Rename instance fields relating to custom default headers

Jitsusama commented:
The requested changes have been made.

lovelydinosaur commented:
Great - let's have it!
Released as 0.6.1
